aips/higher-order-functions.md

```diff
@@ -0,0 +1,132 @@
+in the Move bytecode. This allows them to implement a feature which is
+currently not available for regular Move functions: taking functions,
+given as lambda expressions, as parameters. Given this, we can
+define popular higher-order functions like for_each, filter, map, and fold
+collection types in Move. In this AIP, we suggest a set of conventions
+to add those functions to the data structure in the Move and Aptos frameworks.
+
+## Motivation
+
+It is well-known that higher order functions lead to more concise and correct
+code for collection types. They are widely popular today in mainstream
+languages, including Rust, TypeScript, Java, C++, and more.
+
+## Rationale
+
+Move has currently no traits which would allow to define an `Iterator` type
+which comprehends available functions across multiple collection types.
+Here, we want to establish at least a convention for naming and semantics of
+the most common of such functions. This allows framework writers
+to know which functions to provide, developers to remember
+which functions are available, and auditors to understand what they mean
+in the code.
+
+## Specification
+
+### Foreach
+
+Each iterable collection SHOULD offer the following three functions
+(illustrated by the `vector<T>` type):
+
+```move=
+public inline fun for_each<T>(v: vector<T>, f: |T|);
+public inline fun for_each_ref<T>(v: &vector<T>, f: |&T|);
+public inline fun for_each_mut<T>(v: &mut vector<T>, f: |&mut T|);
+```
+
+Each of those functions iterates over the collection in the
+order specific to that collection. The first one consumes the collection,
+the second one allows to refer to the elements, and the last one to update
+the elements. Here is an example using `for_each_ref`:
+
+```move=
+fun sum(v: &vector<u64>): u64 {
+  let r = 0;
+  for_each_ref(v, |x| r = r + *x);
+  r
+}
+```
+
+### Fold, Map, and Filter
+
+Each iterable collection SHOULD offer the following three functions
+which transpose into a new collection of the same or different type:
+
+```move=
+public inline fun fold<T, R>(v: vector<T>, init: R, f: |R,T|R ): R;
+public inline fun map<T, S>(v: vector<T>, f: |T|S ): vector<S>;
+public inline fun filter<T:drop>(v: vector<T>, p: |&T|bool) ): vector<T>;
+```
+
+To illustrate the semantics of the `fold` and the `map` function, we show
+the definition of the later in terms of the former:
+
+```move=
+public inline fun map<T, S>(v: vector<T>, f: |T|S): vector<S> {
+    let result = vector<S>[];
+    for_each(v, |elem| push_back(&mut result, f(elem)));
+    result
+}
+```
+
+### Notes
+
+- Only collection types which are expected to be iterable in a single
+  transaction should offer these functions. For example, tables are not such
+  collections.
+- Some collection types may diverge by adding less or more of those
+  functions, depending on the type of the collection.
+
+### Effected Data Types in Aptos
+
+Those data types in the Aptos frameworks should get the higher-order functions:
```

Comment:
it might be worth calling out that the expectation that in the future it is an expectation that other data structures will implement them as a prerequisite for inclusion